### DISCUSSION TITLE
runThen() not obeying my "breakThens"?

### DISCUSSION THREAD

**Discussion by migueldilaj** [synfron/ReshaperForBurp#22] [Q&A] [answered]
Hello again!

I may be doing something wrong, so advice welcome.

I've a single rule with this on my Thens:

     RunScript:
        const myLog2 = {text: 'this should BREAK from Thens'};
        Reshaper.event.runThen('Log', myLog2);
        const breakThens = {breaktype:'BreakThens'};
        Reshaper.event.runThen('Break', breakThens);

     Log:
        Text: "I'm past the check on the next THEN!!"

**If I understood it correctly**, the above should log a message telling I should break from Thens, then break execution of next Thens within the same rule.
However, I get this in the logs:

     {
       "message" : "this should BREAK from Thens"
     }
     {
       "message" : "I'm past the check on the next THEN!!"
     }

Is this a bug or an error on my side?
Thank you in advance!

**Comment by ddwightx** [comment, marked as answer]
Hi @migueldilaj. This one is actually a bug, but the bug is that it was a mistake to add Then Break to the list of Thens that could be triggered from scripts because it won't be able to work that way. However, I will add more direct support for a script to trigger a Break at the end of script execution. I will do so via a new function. Thanks for reporting that.

**Comment by migueldilaj** [comment]
Perfect. Thanks a lot for clarifying that!
IMHO, adding support to break further Thens in the same rule, as well as break further Rules, is what provides the real flow control (in the lack of a more specific "If" clause in the Thens).
What is the intended usage for "Continue" in the breakThens()? Just do nothing, or jump to the next Rule? If it's the second, it can be equivalent (for my use) to breakThens, just need to revert the logic.

**Reply by ddwightx** [reply]
Yes, Then Break and Then Run Rule are the "replacement" for the lack of "if" clauses in Thens. 

Continue is a default do nothing, continue as normal case. Under the hood every Then action actually returns a Break value. All of them always return Continue except for Then Break, Run Rule, Run Process, and Send Request based on their context. I will make Then Run Script support this as well, it just can't be done the way (i.e. `Reshaper.event.runThen`) I originally thought it possible.

**Reply by ddwightx** [reply]
@migueldilaj A new API has been added for triggering a break in v1.8.2. See https://synfron.github.io/ReshaperForBurp/ScriptingLibrary.html#setruleresponseruleresponse
